src/bm_sim/lpm_trie.h

```diff
@@ -21,77 +21,79 @@
 #ifndef BM_SIM_LPM_TRIE_H_
 #define BM_SIM_LPM_TRIE_H_
 
-#include <bf_lpm_trie/bf_lpm_trie.h>
+#include <bm/bm_sim/bytecontainer.h>
 
 #include <algorithm>  // for std::swap
-#include <utility>  // for std::swap
+#include <utility>   // for std::swap
+#include <string>
+#include <memory>
+#include <cassert>
+#include <cstdint>
+#include <vector>
 
 namespace bm {
 
-static_assert(sizeof(value_t) == sizeof(uintptr_t),
-              "Invalid type sizes");
+using value_t = std::uintptr_t;
+using byte_t = unsigned char;
+
+static_assert(sizeof(value_t) == sizeof(uintptr_t), "Invalid type sizes");
+
+class Node;
 
 class LPMTrie {
  public:
-  explicit LPMTrie(size_t key_width_bytes)
-    : key_width_bytes(key_width_bytes) {
-    trie = bf_lpm_trie_create(key_width_bytes, true);
-  }
+  explicit LPMTrie(std::size_t key_width_bytes);
 
   /* Copy constructor */
-  LPMTrie(const LPMTrie& other) = delete;
+  LPMTrie(const LPMTrie &other) = delete;
 
   /* Move constructor */
-  LPMTrie(LPMTrie&& other) noexcept
-  : key_width_bytes(other.key_width_bytes), trie(other.trie) {}
+  LPMTrie(LPMTrie &&other) noexcept;
 
-  ~LPMTrie() {
-    bf_lpm_trie_destroy(trie);
-  }
+  ~LPMTrie();
```

Comment:
```suggestion
  ~LPMTrie() = default;
```

and remove the manual implementation. My C++ is rusty but I think that's always better to use `default`?

Reply:
+1

Reply:
since LPMTrie has a unique ptr to Node, it has to know the size of it in destructor, so i put the default impl to .cc file, and have to leave the decl here.

Reply:
We set default in the implementation file otherwise it would create the following build error 

```
/usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/unique_ptr.h:79:16: error: invalid application of 'sizeof' to an incomplete type 'bm::Node'
        static_assert(sizeof(_Tp)>0,
                      ^~~~~~~~~~~
/usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/unique_ptr.h:292:4: note: in instantiation of member function 'std::default_delete<bm::Node>::operator()' requested here
          get_deleter()(std::move(__ptr));
          ^
./lpm_trie.h:46:3: note: in instantiation of member function 'std::unique_ptr<bm::Node, std::default_delete<bm::Node> >::~unique_ptr' requested here
  ~LPMTrie() = default;
  ^
./lpm_trie.h:35:7: note: forward declaration of 'bm::Node'
class Node;
      ^
1 error generated.
make[2]: *** [Makefile:587: lookup_structures.lo] Error 1
```